bazel-build.el

```diff
@@ -0,0 +1,197 @@
+;;; bazel-build.el --- Emacs utilities for using Bazel
+;; Copyright (C) 2018 Google LLC
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;      http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+;;; Commentary:
+;;
+;; This package provides commands to build and run code using Bazel.
+;; It defines interactive commands bazel-build and bazel-run which
+;; perform completion of available Bazel targets.
+
+;;; -*- lexical-binding: t; -*-
+;;; Code:
+
+(require 'cl-lib)
+
+(defun bazel-build ()
+  "Find and build a Bazel target."
+  (interactive)
+  (funcall-interactively #'bazel-build--run-bazel-command "build"))
+
+(defun bazel-run ()
+  "Find and run a Bazel target."
+  (interactive)
+  (funcall-interactively #'bazel-build--run-bazel-command "run"))
+
+(defun bazel-build--run-bazel-command (command)
+  "Run Bazel tool with given COMMAND, e.g. build or run."
+  (let* ((file-name (buffer-file-name))
+         (workspace-root (bazel-build--find-workspace-root file-name))
+         (initial-input (bazel-build--suggest-initial-input file-name workspace-root))
+         (target (completing-read
+                  (format "bazel %s " command) ; prompt
+                  #'bazel-build--completions   ; collection
+                  nil                          ; predicate
+                  nil                          ; require-match
+                  initial-input)))             ; initial-input
+    (with-temp-buffer
+      (compile
+       (format "bazel %s %s" command target)))))
+
+(defun bazel-build--find-workspace-root (file-name)
+  "Find the root of the Bazel workspace containing FILE-NAME."
+  (let ((workspace-root (locate-dominating-file file-name "WORKSPACE")))
+    (if workspace-root (expand-file-name workspace-root)
+      (error "Not in a Bazel workspace"))))
+
+(defun bazel-build--suggest-initial-input (file-name workspace-root)
+  "Suggest initial input to the interactive completion.
+FILE-NAME is the file-name of the current buffer.  WORKSPACE-ROOT is
+the root of the Bazel workspace."
+  (concat "//" (directory-file-name (bazel-build--extract-package-name file-name workspace-root))))
+
+(defun bazel-build--extract-package-name (file-name workspace-root)
+  "Extract the nearest Bazel package for FILE-NAME under WORKSPACE-ROOT."
+  (let ((nearest-build-file (or (bazel-build--find-nearest-build-package file-name) workspace-root)))
+    (file-relative-name nearest-build-file workspace-root)))
+
+(defun bazel-build--find-nearest-build-package (file-name)
```

Comment:
Terminology nit: "package" is only the directory relative to the workspace root. Instead of finding a better name, I'd inline this function - it's quite short.

Reply:
Done. Thanks for the suggestion. I think it looks a bit cleaner now too.